keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,656 @@
+      uid:
+  parameters: [11]
+status:
+  conditions: [12]
+    - Type: {BucketAvailable} [13]
+      Status: [14]
+      Reason: [15]
+      Message: [16]
+```
+
+1. `name`: When created by COSI, the `Bucket` name is generated in this format: _<bucketRequest.namespace>"-"<bucketRequest.name>_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name.
+2. `labels`: added by the controller.  Key’s value should be the provisioner name. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+3. `finalizers`: added by the controller to defer `Bucket` deletion until backend deletion ops succeed.
+4. `provisioner`: The provisioner field as defined in the `BucketClass`.  Used by sidecars to filter `Bucket`s. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".
+5. `releasePolicy`: Prescribes outcome of a Delete events. 
+   - _Retain_: (default) the `Bucket` and its data are preserved. The `Bucket` can potentially be reused.
+   - _Delete_: the bucket and its contents are destroyed.
+> Note: the `Bucket`'s release policy is set to "Retain" as a default. Exercise caution when using the "Delete" release policy as the bucket content will be deleted.
+> Note: a `Bucket` is not deleted if it is bound to any `BucketRequest`s.
+6. `anonymousAccessMode`: a string specifying *uncredentialed* access to the Bucket.  This is applicable to cases where the storage instance or objects are intended to be publicly readable and/or writable. One of:
+   - "private": Default, disallow uncredentialed access to the storage instance.
+   - "publicReadOnly": Read only, uncredentialed users can call ListBucket and GetObject.
+   - "publicWriteOnly": Write only, uncredentialed users can only call PutObject.
+   - "publicReadWrite": Read/Write, same as `ro` with the addition of PutObject being allowed.
+> Note: does not reflect or alter the backing storage instances' ACLs or IAM policies.
+7. `bucketClassName`: Name of the associated bucket class (greenfield only).
+8. `bindings`: an array of `BucketAccess.name`(s). If the list is empty then there are no bindings (accessors) to this `Bucket` instance and the `Bucket` can potentially be deleted.
+9. `protocol`: The protocol the application will use to access the storage instance.
+   - `protocolSignature`: Specifies the protocol targeted by this Bucket instance.  One of:
+     - `azureBlob`: data required to target a provisioned azure container and/or storage account.
+     - `s3`: data required to target a provisioned S3 bucket and/or user.
+     - `gcs`: data required to target a provisioned GCS bucket and/or service account.
+10. `allowedNamespaces`: a copy of the `BucketClass`'s allowed namespaces. Additionally, this list can be mutated by the admin to allow or deny namespaces over the life of the bucket.
```

Comment:
It is, in general, not very common, and definitely 'clumsy' to work with object UIDs. Would it make sense to explicitly state (in this document) why this field is not a set of `Namespace` names, but those UIDs are required?

Reply:
This is intended to be a access restriction mechanism. If a namespace from the `allowedNamespaces` list is deleted and another namespace added by the same name, then the new namespace should not have access to this Bucket.

Reply:
That's what I thought. Will require good UX though to be usable, since AFAIK there's no other place in K8s where one is required to provide UIDs directly.